### PR TITLE
Add CLIP training example script

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

This PR adds an example script for training CLIP-like models. This is inspired from the [Transformers example](https://github.com/huggingface/transformers/tree/main/examples/pytorch/contrastive-image-text).
It solves #167.

- [x] Run the full example on one device (3 epochs over COCO)
- [x] Run the full example on 8 devices
- [x] HMP

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by regisss** [comment]
@skaulintel I tested the script quickly and it seems to work fine. Feel free to try it and let me know if you encounter any issue. I suggest you to follow the README to make sure everything is set properly for running the script.

**Comment by regisss** [comment]
I managed to run the example on 1 and 8 devices. However with HMP I got bad results. I'll take a look to see if I can get a good set of bf16 operators.

**Comment by regisss** [comment]
I managed to get HMP to work well by merging the bf16 ops of RoBERTa and ViT. Here is the configuration I get: https://huggingface.co/Habana/clip/blob/main/gaudi_config.json

File: examples/contrastive-image-text/run_clip.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+
+    # Load image_processor, in this script we only use this to get the mean and std for normalization.
+    image_processor = AutoImageProcessor.from_pretrained(
+        model_args.image_processor_name or model_args.model_name_or_path,
+        cache_dir=model_args.cache_dir,
+        revision=model_args.model_revision,
+        use_auth_token=True if model_args.use_auth_token else None,
+    )
+
+    model = AutoModel.from_pretrained(
```
Comment (libinta) [inline examples/contrastive-image-text/run_clip.py:344]: how does the clip-robert is selected for this case? I didn't find anywhere in transformer git has the map

**Comment by skaulintel** [comment]
@regisss 
I verified the command from the ReadMe works on My end.

However, I tried running the following command adapted from you example in the readme for clipvit.  
`python run_clip.py --output_dir ./clip-vit-finetuned --model_name_or_path /scratch2/skaul/clip-vit-large-patch14/ --data_dir /scratch2/skaul/data/ --dataset_name ydshieh/coco_dataset_script --dataset_config_name=2017 --image_column image_path --caption_column caption --remove_unused_columns=False --do_train  --do_eval --per_device_train_batch_size="16" --per_device_eval_batch_size="16" --learning_rate="5e-5" --warmup_steps="0" --weight_decay 0.1 --overwrite_output_dir --use_habana --use_lazy_mode --use_hpu_graphs --gaudi_config_name /scratch2/skaul/vit/gaudi_config.json --throughput_warmup_steps 2`

The scratch directory is where I have the coco dataset downloaded as well as the clipvit pretrained model and clip gaudi_config. I get this error...
hmp:verbose_mode  False
hmp:opt_level O1
[INFO|trainer.py:518] 2023-02-16 22:38:26,134 >> ***** Running training *****
[INFO|trainer.py:519] 2023-02-16 22:38:26,134 >>   Num examples = 35302
[INFO|trainer.py:520] 2023-02-16 22:38:26,134 >>   Num Epochs = 3
[INFO|trainer.py:521] 2023-02-16 22:38:26,134 >>   Instantaneous batch size per device = 16
[INFO|trainer.py:522] 2023-02-16 22:38:26,134 >>   Total train batch size (w. parallel, distributed & accumulation) = 16
[INFO|trainer.py:523] 2023-02-16 22:38:26,134 >>   Gradient Accumulation steps = 1
[INFO|trainer.py:524] 2023-02-16 22:38:26,134 >>   Total optimization steps = 6621
[INFO|trainer.py:525] 2023-02-16 22:38:26,136 >>   Number of trainable parameters = 427616513
  0%|                                                                                                               | 0/6621 [00:00<?, ?it/s]Traceback (most recent call last):
  File "run_clip.py", line 542, in <module>
    main()
  File "run_clip.py", line 513, in main
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/trainer.py", line 395, in train
    return inner_training_loop(
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/trainer.py", line 667, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/trainer.py", line 2539, in training_step
    loss = self.compute_loss(model, inputs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/trainer.py", line 2571, in compute_loss
    outputs = model(**inputs)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 1125, in forward
    text_outputs = self.text_model(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1212, in _call_impl
    result = forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 712, in forward
    hidden_states = self.embeddings(input_ids=input_ids, position_ids=position_ids)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1212, in _call_impl
    result = forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 230, in forward
    embeddings = inputs_embeds + position_embeddings
  File "/usr/local/lib/python3.8/dist-packages/habana_frameworks/torch/hpex/hmp/utils.py", line 212, in wrapper
    return op(*args_cast, **kwds)
RuntimeError: Incompatible input shapes, broadcast not possible. Tensor1 Size: 768 128 16 Tensor2 Size: 768 77 1
  0%|

File: examples/contrastive-image-text/run_clip.py (hunk 1 above)
Comment (regisss) [reply examples/contrastive-image-text/run_clip.py:344]: The `VisionTextDualEncoderModel` is defined here in Transformers: https://github.com/huggingface/transformers/blob/a8eb4f79f946c5785f0e91b356ce328248916a05/src/transformers/models/vision_text_dual_encoder/modeling_vision_text_dual_encoder.py#L162

Here is the mapping for `AutoModel`: https://github.com/huggingface/transformers/blob/a8eb4f79f946c5785f0e91b356ce328248916a05/src/transformers/models/auto/modeling_auto.py#L51

And then the `config.json` file in the folder where your model is saved contains a `text_config` and a `vision_config` as follows:
```
{
  ...,
  "text_config": {
    "_name_or_path": "roberta-large",
    ...
  },
  "vision_config": {
    "_name_or_path": "openai/clip-vit-large-patch14",
    ...
  }
}
```

**Comment by regisss** [comment]
@skaulintel Can I see the name of the files in `/scratch2/skaul/clip-vit-large-patch14/`? And could you also print here the content of the file called `config.json` in `/scratch2/skaul/clip-vit-large-patch14/`?

**Comment by skaulintel** [comment]
> @skaulintel Can I see the name of the files in `/scratch2/skaul/clip-vit-large-patch14/`? And could you also print here the content of the file called `config.json` in `/scratch2/skaul/clip-vit-large-patch14/`?

```
skaul@sc09super14-klb:~$ ls /scratch2/skaul/clip-vit-large-patch14/
config.json         preprocessor_config.json  special_tokens_map.json  tokenizer.json
flax_model.msgpack  pytorch_model.bin         tf_model.h5              vocab.json
merges.txt          README.md                 tokenizer_config.json
```

config.json
`{
  "_name_or_path": "clip-vit-large-patch14/",
  "architectures": [
    "CLIPModel"
  ],
  "initializer_factor": 1.0,
  "logit_scale_init_value": 2.6592,
  "model_type": "clip",
  "projection_dim": 768,
  "text_config": {
    "_name_or_path": "",
    "add_cross_attention": false,
    "architectures": null,
    "attention_dropout": 0.0,
    "bad_words_ids": null,
    "bos_token_id": 0,
    "chunk_size_feed_forward": 0,
    "cross_attention_hidden_size": null,
    "decoder_start_token_id": null,
    "diversity_penalty": 0.0,
    "do_sample": false,
    "dropout": 0.0,
    "early_stopping": false,
    "encoder_no_repeat_ngram_size": 0,
    "eos_token_id": 2,
    "finetuning_task": null,
    "forced_bos_token_id": null,
    "forced_eos_token_id": null,
    "hidden_act": "quick_gelu",
    "hidden_size": 768,
    "id2label": {
      "0": "LABEL_0",
      "1": "LABEL_1"
    },
    "initializer_factor": 1.0,
    "initializer_range": 0.02,
    "intermediate_size": 3072,
    "is_decoder": false,
    "is_encoder_decoder": false,
    "label2id": {
      "LABEL_0": 0,
      "LABEL_1": 1
    },
    "layer_norm_eps": 1e-05,
    "length_penalty": 1.0,
    "max_length": 20,
    "max_position_embeddings": 77,
    "min_length": 0,
    "model_type": "clip_text_model",
    "no_repeat_ngram_size": 0,
    "num_attention_heads": 12,
    "num_beam_groups": 1,
    "num_beams": 1,
    "num_hidden_layers": 12,
    "num_return_sequences": 1,
    "output_attentions": false,
    "output_hidden_states": false,
    "output_scores": false,
    "pad_token_id": 1,
    "prefix": null,
    "problem_type": null,
    "projection_dim" : 768,
    "pruned_heads": {},
    "remove_invalid_values": false,
    "repetition_penalty": 1.0,
    "return_dict": true,
    "return_dict_in_generate": false,
    "sep_token_id": null,
    "task_specific_params": null,
    "temperature": 1.0,
    "tie_encoder_decoder": false,
    "tie_word_embeddings": true,
    "tokenizer_class": null,
    "top_k": 50,
    "top_p": 1.0,
    "torch_dtype": null,
    "torchscript": false,
    "transformers_version": "4.16.0.dev0",
    "use_bfloat16": false,
    "vocab_size": 49408
  },
  "text_config_dict": {
    "hidden_size": 768,
    "intermediate_size": 3072,
    "num_attention_heads": 12,
    "num_hidden_layers": 12,
    "projection_dim": 768
  },
  "torch_dtype": "float32",
  "transformers_version": null,
  "vision_config": {
    "_name_or_path": "",
    "add_cross_attention": false,
    "architectures": null,
    "attention_dropout": 0.0,
    "bad_words_ids": null,
    "bos_token_id": null,
    "chunk_size_feed_forward": 0,
    "cross_attention_hidden_size": null,
    "decoder_start_token_id": null,
    "diversity_penalty": 0.0,
    "do_sample": false,
    "dropout": 0.0,
    "early_stopping": false,
    "encoder_no_repeat_ngram_size": 0,
    "eos_token_id": null,
    "finetuning_task": null,
    "forced_bos_token_id": null,
    "forced_eos_token_id": null,
    "hidden_act": "quick_gelu",
    "hidden_size": 1024,
    "id2label": {
      "0": "LABEL_0",
      "1": "LABEL_1"
    },
    "image_size": 224,
    "initializer_factor": 1.0,
    "initializer_range": 0.02,
    "intermediate_size": 4096,
    "is_decoder": false,
    "is_encoder_decoder": false,
    "label2id": {
      "LABEL_0": 0,
      "LABEL_1": 1
    },
    "layer_norm_eps": 1e-05,
    "length_penalty": 1.0,
    "max_length": 20,
    "min_length": 0,
    "model_type": "clip_vision_model",
    "no_repeat_ngram_size": 0,
    "num_attention_heads": 16,
    "num_beam_groups": 1,
    "num_beams": 1,
    "num_hidden_layers": 24,
    "num_return_sequences": 1,
    "output_attentions": false,
    "output_hidden_states": false,
    "output_scores": false,
    "pad_token_id": null,
    "patch_size": 14,
    "prefix": null,
    "problem_type": null,
    "projection_dim" : 768,
    "pruned_heads": {},
    "remove_invalid_values": false,
    "repetition_penalty": 1.0,
    "return_dict": true,
    "return_dict_in_generate": false,
    "sep_token_id": null,
    "task_specific_params": null,
    "temperature": 1.0,
    "tie_encoder_decoder": false,
    "tie_word_embeddings": true,
    "tokenizer_class": null,
    "top_k": 50,
    "top_p": 1.0,
    "torch_dtype": null,
    "torchscript": false,
    "transformers_version": "4.16.0.dev0",
    "use_bfloat16": false
  },
  "vision_config_dict": {
    "hidden_size": 1024,
    "intermediate_size": 4096,
    "num_attention_heads": 16,
    "num_hidden_layers": 24,
    "patch_size": 14,
    "projection_dim": 768
  }
}
`

**Comment by regisss** [comment]
@skaulintel This looks right. I just tested in Transformers and it fails too so I guess this issue is not related to Optimum Habana.
Let me see why it fails on Transformers' side and I'll follow up.

**Comment by regisss** [comment]
@skaulintel Okay I understand better what is going on. This is not a bug.
Basically, the model `openai/clip-vit-large-patch14` has a config parameter `max_position_embeddings` equal to 77 (see [here](https://huggingface.co/openai/clip-vit-large-patch14/blob/8d052a0f05efbaefbc9e8786ba291cfdf93e5bff/config.json#L47)). Because it was pre-trained on captions that contains maximum 77 tokens. However, in the example `run_clip.py`, captions are by default padded to 128 tokens. Hence the dimension issue.
So I recommend you create a new model with another text encoder (e.g. RoBERTa as shown in the README).
Or if you absolutely want to run this model, add the argument `--max_seq_length 77` to your command.

**Review by libinta — COMMENTED** [review]
@regisss  are we using the same set of bf16 ops as default?

**Comment by regisss** [comment]
> @regisss are we using the same set of bf16 ops as default?

@libinta We use the bf16 ops defined here: https://huggingface.co/Habana/clip/blob/main/gaudi_config.json
Basically I just merged the bf16 ops of ViT with the ones of RoBERTa and that's what I got.

**Comment by regisss** [comment]
@libinta 

I ran a quick benchmark Gaudi vs V100:
- Gaudi 8x bf16
  - batch size: 96
  - throughput: 850,39 samples/s
  - evaluation loss: 1.7529
- V100 8x fp16
  - batch size: 96
  - throughput: 637,259 samples/s
  - evaluation loss: 1.759
  
I also tried running a CLIP model with Swin as the vision encoder and BERT as the text encoder and it works.

**Review by libinta — APPROVED** [review]

**Pull request merged**